### PR TITLE
add documentation for verify-api-groups

### LINKED ISSUE #5532: Document remaining kubernetes/kubernetes hack/verify* tests
Part of developer guide audit for sig-testing: https://github.com/kubernetes/community/issues/5239

Follow-on to https://github.com/kubernetes/community/pull/5349 (ref: https://github.com/kubernetes/community/pull/5349#discussion_r575029064)

We lack documentation for the sundry verify tests that are used to block merge or release of kubernetes/kubernbetes.  We should ensure these tests are documented in [contributors/devel/sig-testing/verify-tests.md](https://git.k8s.io/community/contributors/devel/sig-testing/verify-tests.md).

The following merge-blocking verify tests are run as part of `pull-kubernetes-verify`
- [x] verify-api-groups (https://github.com/kubernetes/community/pull/5829, thanks @pravarag)
- [x] verify-bazel (https://github.com/kubernetes/community/pull/5864, @pravarg)
- [x] verify-boilerplate (https://github.com/kubernetes/community/pull/5864, @pravarag)
- [x] verify-cli-conventions (https://github.com/kubernetes/community/pull/5864, @pravarag)
- [x] verify-codegen (https://github.com/kubernetes/community/pull/5864, @pravarag)
- [ ] verify-conformance-requirements
- [ ] verify-description
- [ ] verify-flags-underscore
- [ ] verify-generated-device-plugin
- [ ] verify-generated-docs
- [ ] verify-generated-files
- [ ] verify-generated-kms
- [ ] verify-generated-kubelet-plugin-registration
- [ ] verify-generated-pod-resources
- [ ] verify-generated-protobuf
- [ ] verify-generated-runtime
- [ ] verify-generated-stable-metrics
- [ ] verify-generated-swagger-docs
- [ ] verify-gofmt
- [ ] verify-govet
- [ ] verify-import-aliases
- [ ] verify-import-boss
- [ ] verify-imports
- [ ] verify-internal-modules
- [ ] verify-no-vendor-cycles
- [ ] verify-openapi-spec
- [ ] verify-pkg-names
- [ ] verify-prerelease-lifecycle-tags
- [ ] verify-publishing-bot
- [ ] verify-readonly-packages
- [ ] verify-shellcheck
- [ ] verify-spelling
- [ ] verify-staging-meta-files
- [ ] verify-staticcheck
- [ ] verify-test-code
- [ ] verify-test-featuregates
- [ ] verify-test-images

There are also additional tests that are run as part of release-blocking verify jobs:
- [ ] verify-external-dependencies-version
- [ ] verify-generated-files-remake
- [ ] verify-readonly-packages
- [ ] verify-typecheck
- [ ] verify-typecheck-dockerless
- [ ] verify-typecheck-providerless
- [ ] verify-vendor
- [ ] verify-vendor-licenses

/priority important-longterm
/kind documentation
/sig testing
/help

### PR DESCRIPTION
**Pull request by pravarag**
**Which issue(s) this PR fixes**:

Fixes #5532

As part of this PR, documentation for test `verify-api-groups` have been added.

/sig testing

### CODE REVIEW

File: contributors/devel/sig-testing/verify-tests.md [hunk 1]
```diff
@@ -81,3 +82,44 @@ In order of decreasing preference:
...
 Contact SIG-Security with any additional questions.
+
+### `verify-api-groups`
+
+This verification script validates the different api-groups by reading
+the respective `register.go` file. Every register file should contain
+a GroupName from where the different representations can be
+gathered. Also, the other representations will be, `group_dirnames`,
+`external_group_versions`, `expected_install_packages`.
+
```
Comment (spiffxp) [inline contributors/devel/sig-testing/verify-tests.md:93]: Should implies convention, use 'must' instead if it's a hard requirement.  Remove implementation details and extra blank line
```suggestion
the respective `register.go` file. Every register file must contain
a `GroupName`.
```

File: contributors/devel/sig-testing/verify-tests.md [hunk 2]
```diff
@@ -81,3 +82,44 @@ In order of decreasing preference:
...
+```bash
+if grep -q 'GroupName = "' "${register_file}"; then
+		group_name=$(grep 'GroupName = "' "${register_file}" | cut -d\" -f2 -)
+else
+		echo "${register_file} is missing \"const GroupName =\""
+		exit 1
+fi
+```
+
```
Comment (spiffxp) [inline contributors/devel/sig-testing/verify-tests.md:109]: Remove implementation detail
```suggestion
```

File: contributors/devel/sig-testing/verify-tests.md [hunk 3]
```diff
@@ -81,3 +82,44 @@ In order of decreasing preference:
...
+Another check which is performed when this script runs is to ensure
+client gen is being collected properly. There is a list called
+`groups_without_codegen` which holds name of groups we expect not to
+have clients generated for them. And this happens for types that
+aren't served from API server.
```
Comment (spiffxp) [inline contributors/devel/sig-testing/verify-tests.md:114]:
```suggestion
Another check which is performed when this script runs is to ensure
that all types have client code generated for them, except types that
belong to groups not served from the API server (defined in this 
script via the bash array `groups_without_codegen`)
```

File: contributors/devel/sig-testing/verify-tests.md [hunk 4]
```diff
@@ -81,3 +82,44 @@ In order of decreasing preference:
...
+
+In the last check of the script, we check for `import_known_versions`
+to ensure the import packages will get installed. We list out packages
+which are required without instllation along with importing
```
Comment (spiffxp) [inline contributors/devel/sig-testing/verify-tests.md:118]:
```suggestion
which are required without installation along with importing
```

File: contributors/devel/sig-testing/verify-tests.md [hunk 5]
```diff
@@ -81,3 +82,44 @@ In order of decreasing preference:
...
+fi
+```
+
+Another check which is performed when this script runs is to ensure
+client gen is being collected properly. There is a list called
+`groups_without_codegen` which holds name of groups we expect not to
+have clients generated for them. And this happens for types that
+aren't served from API server.
+
+In the last check of the script, we check for `import_known_versions`
```
Comment (spiffxp) [inline contributors/devel/sig-testing/verify-tests.md:116]:
```suggestion
Next, the script compares the `GroupName`s against `import_known_versions.go`
```

File: contributors/devel/sig-testing/verify-tests.md [hunk 6]
```diff
@@ -81,3 +82,44 @@ In order of decreasing preference:
...
+to ensure the import packages will get installed. We list out packages
+which are required without instllation along with importing
+`known_version`. Then we do a search for packages that reqiure
+installation on the basis of `packages_without_installation`. We
+verify if file is a `known_version_file` or not only if an
+`expected_install_package` is present in it.
+
+At the end, we do a check of all the `groupversions` to ensure they
+are in the init.sh file.
```
Comment (spiffxp) [inline contributors/devel/sig-testing/verify-tests.md:125]:
```suggestion
Finally the script checks that all external group versions (e.g. `foobar/v1`) are defined
in `hack/lib/init.sh` in either the `KUBE_AVAILABLE_GROUP_VERSIONS` or
`KUBE_NONSERVER_GROUP_VERSIONS` bash variables
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
Thank you for helping out!  Some suggestions to try and reduce the amount of text someone needs to read to answer "what is this verifying?"

File: contributors/devel/sig-testing/verify-tests.md (hunk 6 above)
Comment (pravarag) [reply contributors/devel/sig-testing/verify-tests.md]: Updated as per suggestion.

File: contributors/devel/sig-testing/verify-tests.md (hunk 3 above)
Comment (pravarag) [reply contributors/devel/sig-testing/verify-tests.md]: Updated as per suggestions

**Comment by pravarag** [comment]
I've updated the code as suggested in review comments.

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Pull request merged**